keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1067 @@
+  option (e.g. via a controller).
+- CRD lifecycle prevents CRs from being written until the CRD is written, which
+  means that until a ValidatingAdmissionConfiguration is written AND a CRD is
+  created for it, no CRs can be written.  Applying multiple resources
+  (e.g. kubectl apply -f <lots of yaml files>) only works if the apply
+  operations are ordered and objects are verified to exist at specific steps.
+
+#### Additional Capabilities (WIP)
+
+In addition to the high level API structure proposed above, there are quite
+a few capatiblities that are important to this enhancement that we are working
+on defining. These are all work-in-progress.
+
+Rule Message:
+
+- message formatting (need support for string formatting. Allow a CEL expression
+  to be used to format the whole string? Or allow CEL expressions to evaluate
+  template values?)
+- severity (e.g. "Warning")
+- status type equivalent (e.g. Forbidden can be returned from a webhook using
+  HTTP 403) (can a match rule override this? or can it be selected via a CEL
+  expression?)
+- Similar to how each match has configuration, can each match also select
+  the severity of messages?
+
+Rule scope:
+
+- Dereferencing to deeply nested fields in CEL can be burdonsome. A field
+  containing a field path would allow a CEL rule to be scope to some node in
+  the schema.
+- Also, the idea of named scopes (e.g. "ALL_CONTAINERS") would allow a single
+  rule to be applied to all container definitions in a resource (e.g. both
+  "containers" and "initContainers" in a pod)
+
+Failure Policy:
+
+- Not the same as webhooks, where failure policy is about the remote request
+- What if a CEL rule evaluates to an error?
+- What if a CEL rule is for a CRD kind and the CRD changes such that the CEL
+  rule fails type checking? (or the CRD does not exist?)
+
+Decision Rules:
+
+- Are overlapping matches combined using an OR or AND? Can this be configured?
+- Can a validation rule depend on another validation rule (only evaluate if the
+  other rule is true?)
+- How to make it so that an "not authorized" error prevents all other validation
+  messages from being returned (to prevent exfiltration)?
+
+Composition utilities:
+
+- Ability to define a sub-expression and then use it in validation rules (xref
+  cel-policy-template "terms")
+
+Transition rules:
+
+- Should we use "self" and "oldSelf" ? How do transition rules work for delete?
+
+Access to namespace labels:
+
+- Most heavily needed fields not directly available in the resource being
+  validated. Note that namespaceSelectors already allow matches to examine
+  namespace levels.
+
+Type safety:
+
+- What happens when a ValidatingAdmissionConfiguration is created in a cluster
+  where either CRDs or native types are incompatible with CEL rules and the CEL
+  rules fail to compile?
+  - Option: Put something into status about the type check error and apply the
+    fail policy instead of evaluating the CEL expression at runtime
```

Comment:
status on API server configuration objects has been tricky to design in the past, because of the following:
* multiple active kube-apiservers (sometimes at identical versions, sometimes skewed by one version during upgrade)
* multiple active non-kube-apiserver servers (aggregated servers)

Reply:
Looks 'NonStructural' relies on the fact that the status change is a function of the resource then generation can be used (https://github.com/kubernetes/kubernetes/commit/2cfc3c69dc7c17b2711af0168f39ed7f515675c2). That's not the case here, we want to have both the CEL rules (defined in one resource) and the CRD schema defined in another resource (or native type schema, which is version dependent).

I'll make note of this in the KEP.